server/src/main/java/org/elasticsearch/indices/recovery/plan/ShardRecoveryPlanner.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.indices.recovery.plan;
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.store.Store;
+import org.elasticsearch.index.store.StoreFileMetadata;
+
+import java.util.List;
+
+public interface ShardRecoveryPlanner {
+    @Nullable
+    ShardRecoveryPlan computePlan(String shardIdentifier,
+                                  String shardHistoryUUID,
+                                  Store.MetadataSnapshot sourceMetadata,
+                                  Store.MetadataSnapshot targetMetadata,
+                                  long startingSeqNo,
+                                  int translogOps,
+                                  Store.RecoveryDiff sourceTargetDiff,
```

Comment:
We could simplify this method signature a bit if we let each planner to compute `sourceTargetDiff` and `filesMissingInTarget` but it seems a bit wasteful?